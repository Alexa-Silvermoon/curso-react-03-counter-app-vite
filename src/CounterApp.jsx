import PropTypes from 'prop-types';

const handleAdd = ( event, newValue ) => {

    console.log( event );
    console.log( newValue );

}

// se crea rapido escribiendo rafc
export const CounterApp = ( { value } ) => {
    //<CounterApp value={ 123456 }></CounterApp>

    /* function handleAdd( event, newValue ){

        console.log( event );
        console.log( newValue );
    } 
    */

    /* const handleAdd = ( event, newValue ) => {

        console.log( event );
        console.log( newValue );
    
    } 
    */

    return(

        <>
            <h1>CounterApp - h1</h1>
            <h2> { value } </h2>

            <button onClick={ /* ( event ) => */ handleAdd/* ( event ) */ }> +1 </button>
            <button onClick={ ( event ) => handleAdd( event, 'hola chaval' ) }> +1 </button>
        </>
    )
}

CounterApp.propTypes = {

    value: PropTypes.number.isRequired //define value como numero y es siempre requerido un numero ahi
}