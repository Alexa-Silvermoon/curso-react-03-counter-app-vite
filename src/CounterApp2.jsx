
//USO DE HOOKS CON useState

import { useState } from 'react';//useState es una funcion que devuelve un arreglo

// se crea rapido escribiendo rafc
export const CounterApp2 = ( { valorInicial } ) => {

    console.log( 'render' );
    //API.. FETCH...

    const [ counter, setCounter ] = useState( valorInicial ); // 0 es el valor inicial
    // el setCounter le dice a React que el estado cambio, asi que hay que renderizar el componente de nuevo

    const aumentar = (  ) => {

        setCounter( counter + 1 ); // forma 1
        // setCounter( ( counter ) =>  counter + 1  ); // forma 2

    }
    
    const reducir = (  ) => {

        setCounter( counter - 1 );

    }
    const resetear = (  ) => {

        // setCounter( counter - counter ); // tecnicamente no es valido pero si funciona
        setCounter( valorInicial );

    }

    return(

        <>
            <h1>CounterApp2 - h1</h1>
            <h2> { counter } </h2>

            <button onClick={ /* ( event ) => */ aumentar/* ( event ) */ }> +1 </button>
            <button onClick={ reducir}> -1 </button>
            <button onClick={ resetear }> resetear </button>
        </>
    )
}

//NOTA: el useState hook hace que el valor no sea mutable con el props