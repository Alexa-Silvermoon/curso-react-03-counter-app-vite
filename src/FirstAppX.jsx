
import PropTypes from 'prop-types'; //hace obligatorio lo que venga desde main.jsx, ya sea string o number

export const FirstAppX = ( { title, subTitle } ) => {

    return(

        <>
        
            {/* <div data-testid="test-title">{title}</div> */} {/* esta linea sirve para documento FirstAppX.test.jsx */}
            <h1 data-testid="test-title">{title}</h1> {/* esta linea sirve para documento FirstAppX.test.jsx */}
            <p>{subTitle}</p>
            <p>{subTitle}</p>
            {/* <h1>{title}</h1> */}
            <p>p en FirstAppX.jsx</p>
        
        </>
        
    );

}

FirstAppX.propTypes = {

    title: PropTypes.string.isRequired,
    //hace que lo que venga en el title sea obligatoriamente un string, sino da error

    subTitle: PropTypes.string

}