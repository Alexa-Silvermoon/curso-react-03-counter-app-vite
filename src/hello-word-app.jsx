// import { Fragment } from "react";// es para usar multiples etiquetas en el return

import PropTypes from 'prop-types'; //hace obligatorio lo que venga desde main.jsx, ya sea string o number
// en caso de no tener instalado prop-types, abrir cmd y ejecutar comando
// yarn add prop-types

export function App2() {

    return (<h1>hola mundo desde App2 hello-word-app.jsx</h1>);
}

export const App3 = () => {

    return(
        <h1>hola mundo desde App3 hello-word-app.jsx</h1>
    )
}

export const FirstApp = () => {

    return(
        <h1> Christian Alexander Martinez Millan </h1>
    );

}

const nuevoMensaje = 'mensaje como const'
const arreglo = [1,2,3,4,5];

const mensajeObjeto = {

    mensaje: 'mensaje objeto',
    titulo: 'mensaje objeto - titulo'
}

const getResult = (a,b) => {

    const vocales = ['a','e','i','o','u'];
    const numeros = [1,2,3,4,5,6,7,8,9];
    const c = a + b;
    
    const fusion = c + vocales + numeros ;
    return( fusion );
}

export const FirstApp2 = () => {

    return( //en caso de no querer usar fragment, se puede usar div, o se puede usar <></> pero vacios, esto se le conoce como nodo padre

        <>
            <h1> Christian Alexander Martinez Millan </h1>
            <p> estoy estudiando react + vite </p>
            <h1> { nuevoMensaje } </h1>
            <h1> { arreglo } </h1>
            <h1> { JSON.stringify( mensajeObjeto ) } </h1>
            <h1> { getResult(2,2) } </h1>
        </>
    );

    /* 
    return(

        <Fragment>
            <h1> Christian Alexander Martinez Millan - Fragment </h1>
            <p> estoy estudiando react + vite </p>
        </Fragment>
    );
     */

     /* 
    return(

        <div>
            <h1> Christian Alexander Martinez Millan </h1>
            <p> estoy estudiando react + vite </p>
        </div>
    );
     */

}

// export const FirstApp3 = ( props ) => {
// export const FirstApp3 = ( { title = 'se te esta olvidando el title mi rey' } ) => { //desestructure el props
export const FirstApp3 = ( { title, subTitle, subTitle2, name } ) => { //desestructure el props
    
// export const FirstApp3 = ( { title, subTitle, subTitle2 = 'la tierra es plana' } ) => { //desestructure el props

    //         <FirstApp3 title="soy el title" subTitle="soy el subTitle"></FirstApp3>
    // comilla doble lo convierte en string

    /* 
    en componenets en consola de chrome:
    {
  "title": "hola soy goku"
    }
    */

    // console.log( props );

    return(

        <>
            {/* <h1>{ props.title }</h1> */}
            <h1>{ title }</h1>
            <p> { subTitle + " alexander " + 95 } </p>   {/* 95 se vuelve string */}
            <p> { subTitle2 }</p> {/* 1995 si es un numero */}
            <p> soy un p </p>
            <p> { name } </p>
        </>
    )
}

FirstApp3.propTypes = {

    title: PropTypes.string.isRequired,
    //hace que lo que venga en el title sea obligatoriamente un string, sino da error

    subTitle2: PropTypes.number
    //hace que sea un numero obligatoriamente, sino da error

}

FirstApp3.defaultProps = {//defaultProps se activan antes que los propTypes

    title: 'ho hay titulo - defaultProps',//parece esto en caso de que title="soy el title" no exista en main.jsx
    subTitle: 'no hay subtitulo - defaultProps',
    name: 'Alexander - defaultProps' //no existe en main.jsx pero igual aparece en el html gracias a los defaultProps
}