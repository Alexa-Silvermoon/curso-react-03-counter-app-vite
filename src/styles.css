html, body {
    background-color: #21232A;
    color: white;
    font-family: Helvetica, Arial , sans-serif;
    font-size: 1.3rem;
    padding: 70px;
}

/* para  ordenar alfabaticamente Ctrl + Shift + P    y seleccionar Sort Lines Ascending o Descending*/

button{

    padding: 15px;
    font-size: 1.3rem;
    margin-right: 5px;
}