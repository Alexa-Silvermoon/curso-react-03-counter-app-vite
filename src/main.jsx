// escribir imr para escribirlo rapidamente:
import React from 'react';
import ReactDOM from 'react-dom/client'; //para renderizar
import { FirstAppX } from './FirstAppX';
import { CounterApp } from './CounterApp';
import { CounterApp2 } from './CounterApp2';
import { App2, App3, FirstApp, FirstApp2, FirstApp3 } from './hello-word-app'; //PascalCase en react para los componenetes la primera letra debe ser en mayusculas
import './styles.css';

function App() {

    return (<h1>hola mundo desde main.jsx</h1>);
}

ReactDOM.createRoot( document.getElementById('root')).render(

    <React.StrictMode> {/* este bloque es conocido como el padre */}
        <CounterApp value={100}></CounterApp>
        <CounterApp2 valorInicial={ 0 }></CounterApp2>
        <App></App>
        <App2></App2>
        <App3></App3>
        <FirstApp></FirstApp>
        <FirstApp2></FirstApp2>
        <FirstApp3 title="soy el title" subTitle="soy el subTitle" subTitle2={ 1995 }></FirstApp3>
        <FirstAppX title="title en FirstAppX - main.jsx" subTitle="subtitulo en FirstAppX"></FirstAppX>
    </React.StrictMode>
)

// https://www.udemy.com/course/react-cero-experto/learn/lecture/19663702#questions
