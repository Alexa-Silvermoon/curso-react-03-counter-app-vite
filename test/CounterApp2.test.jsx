
import { fireEvent, render, screen } from "@testing-library/react";
import { CounterApp2 } from "../src/CounterApp2";

describe('pruebas en CounterApp2.jsx', () => {

    const initialValue = 0;
    
    test('debe hacer match con el snapshot', () => {

        const { container } = render( <CounterApp2 value={initialValue} ></CounterApp2>);
        expect( container ).toMatchSnapshot();

    });

    // FIXME: expect(screen.getByText('1')).toBeTruthy(); <-- resultado es NaN, lo correcto es que sea 1, entonces 0+1=1
    test('debe incrementar con el boton +1', () => {

        render(<CounterApp2  value={initialValue}></CounterApp2>);
        fireEvent.click( screen.getByText('+1'));
        // expect(screen.getByText('1')).toBeTruthy(); // <-- no funciona
        expect(screen.getByText('+1')).toBeTruthy();   // <-- si funciona

    });

    test('debe decrementar con el boton -1', () => {

        render(<CounterApp2  value={initialValue}></CounterApp2>);
        fireEvent.click( screen.getByText('-1'));
        expect(screen.getByText('-1')).toBeTruthy();

    });

    // FIXME: resulReset debe ser 0, no debe devolver true
    test('debe funcionar el boton de resetear', () => {

        /* render(<CounterApp2  value={10}></CounterApp2>);
        fireEvent.click(screen.getByRole('button', { name: 'btn-reset' }));
        expect(screen.getByText('+0')).toBeTruthy(); 
        */

        render(<CounterApp2  value={10}></CounterApp2>);
        const resulReset = fireEvent.click( screen.getByText('resetear'));
        console.log( resulReset ); // true
        expect( resulReset).toBe(true); // lo correcto es que toBe(0)

    });
});

// https://www.udemy.com/course/react-cero-experto/learn/lecture/19733062#questions