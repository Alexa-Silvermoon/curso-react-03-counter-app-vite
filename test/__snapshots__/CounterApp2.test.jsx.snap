// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas en CounterApp2.jsx debe hacer match con el snapshot 1`] = `
<div>
  <h1>
    CounterApp2 - h1
  </h1>
  <h2>
     
     
  </h2>
  <button>
     +1 
  </button>
  <button>
     -1 
  </button>
  <button>
     resetear 
  </button>
</div>
`;
